src/viam/sdk/components/encoder/encoder.hpp

```diff
@@ -0,0 +1,114 @@
+#include <viam/sdk/registry/registry.hpp>
+#include <viam/sdk/resource/resource_manager.hpp>
+
+namespace viam {
+namespace sdk {
+
+/// @defgroup Encoder Classes related to the `Encoder` component.
+
+/// @class EncoderSubtype
+/// @brief Defines a `ResourceSubtype` for the `Encoder` component.
+/// @ingroup Encoder
+class EncoderSubtype : public ResourceSubtype {
+   public:
+    std::shared_ptr<ResourceServerBase> create_resource_server(
+        std::shared_ptr<ResourceManager> manager) override;
+    std::shared_ptr<ResourceBase> create_rpc_client(std::string name,
+                                                    std::shared_ptr<grpc::Channel> chan) override;
+    EncoderSubtype(const google::protobuf::ServiceDescriptor* service_descriptor)
+        : ResourceSubtype(service_descriptor){};
+};
+
+/// @class Encoder encoder.hpp "components/encoder/encoder.hpp"
+/// @brief An encoder is a device that is hooked up to motors to report a position
+/// @ingroup Encoder
+///
+/// This acts as an abstract base class to be inherited from by any drivers representing
+/// specific encoder implementations. This class cannot be used on its own.
+class Encoder : public ComponentBase {
+   public:
+    /// @enum position_type
+    enum position_type {
+        // Unspecified position type
+        UNSPECIFIED,
+        // Provided by incremental encoders
+        TICKS_COUNT,
+        // Provided by absolute encoders
+        ANGLE_DEGREES
+    };
+
+    /// @struct position
+    /// @brief reported position.
+    struct position {
+        float value;
+        position_type type;
+    };
+
+    /// @struct properties
+    /// @brief Encodes the supported modes of this encoder
+    struct properties {
+        bool ticks_count_supported;
+        bool angle_degrees_supported;
+    };
+
+    // functions shared across all components
+    static std::shared_ptr<ResourceSubtype> resource_subtype();
+    static Subtype subtype();
+
+    /// @brief Creates a `properties` struct from its proto representation.
+    static position_type from_proto(viam::component::encoder::v1::PositionType proto);
+
+    /// @brief Creates a `position` struct from its proto representation.
+    static position from_proto(viam::component::encoder::v1::GetPositionResponse proto);
+
+    /// @brief Creates a `properties` struct from its proto representation.
+    static properties from_proto(viam::component::encoder::v1::GetPropertiesResponse proto);
+
+    /// @brief Converts a `position` struct to its proto representation.
+    static viam::component::encoder::v1::PositionType to_proto(position_type position_type);
+
+    /// @brief Converts a `position` struct to its proto representation.
+    static viam::component::encoder::v1::GetPositionResponse to_proto(position position);
+
+    /// @brief Converts a `properties` struct to its proto representation.
+    static viam::component::encoder::v1::GetPropertiesResponse to_proto(properties properties);
```

Comment:
Are these a necessary part of the `Encoder` class API? Could they be declared/defined somewhere that isn't public and made available only to the `EncoderClient` and `EncoderServer` implementations?

Reply:
That is a good question. I had assumed that this was done this way so that if we are late to implementing part of the API, customers are able to quickly patch/build it themselves. However that may not be what we want.

Reply:
Let's keep it for now. So far we haven't made a decision either way about whether the `proto` types should form part of the public API for the SDK. Currently they do, but we also write wrapper types (as you are doing with e.g. `struct position` above), so we sort of have a foot in each camp. I'm not sure which way we will resolve it, but it won't happen in this review. I think we can leave them for now but something to think about going forward.